### PR TITLE
Engine rewrite

### PR DESCRIPTION
**Pull request by danhp**

### CODE REVIEW

File: Warfare/Warfare/GameEngine.swift [hunk 1]
```diff
@@ -17,6 +17,331 @@ class GameEngine {
...
+    // Set up the gameState after which the player can start giving out orders
+    func beginTurn() {
+        for village in self.game.currentPlayer.villages {
+            for tile in village.controlledTiles {
+
+                // Replace tombstones
+                tile.replaceTombstone()
+
+                // Produce constructions and set unit actions.
+                // ReadyForOrders for all except first phase cultivation
```
Comment (justindomingue) [inline Warfare/Warfare/GameEngine.swift:67]: what about a master "if" : `if let action = tile.unit?.currentAction { }` and then use action?

File: Warfare/Warfare/GameEngine.swift [hunk 2]
```diff
@@ -17,6 +17,331 @@ class GameEngine {
...
+            tile.structure = nil
+            if tile.unit != nil {
+                tile.unit = nil
+                tile.structure = .Tombstone
+            }
+            tile.village = nil
+            tile.owner.removeTile(tile)
+            self.game.neutralTiles.append(tile)
+        }
+        self.game.currentPlayer.removeVillage(village)
```
Comment (justindomingue) [inline Warfare/Warfare/GameEngine.swift]: doesn't removeVillage also call tile.clear for each tile controlled by the village?

File: Warfare/Warfare/GameEngine.swift [hunk 3]
```diff
@@ -17,6 +17,331 @@ class GameEngine {
...
+    }
+
+    func moveUnit(from: Tile, to: Tile) {
+        if from.owner.player !== self.game.currentPlayer { return }
+
+        var path = [Tile]()
+        var village = from.owner
```
Comment (justindomingue) [inline Warfare/Warfare/GameEngine.swift]: could be a constant

File: Warfare/Warfare/GameViewController.swift [hunk 4]
```diff
@@ -49,7 +49,7 @@ class GameViewController: UIViewController {
 		if tileSelected?.owner == nil || tileSelected?.owner.player !== GameEngine.Instance.game.currentPlayer { return }
 
         if let t = tileSelected?.owner {
-            GameEngine.Instance.game?.recruitUnit(t, type: Constants.Types.Unit.Peasant, tile: tileSelected!)
+			GameEngine.Instance.recruitUnit(tileSelected!, type: Constants.Types.Unit.Peasant)
```
Comment (justindomingue) [inline Warfare/Warfare/GameViewController.swift]: oops my bad here - the `if let t...` is rather useless

File: Warfare/Warfare/Map.swift [hunk 5]
```diff
@@ -156,7 +156,7 @@ class Map: SKNode {
 
     func getVillage(region: [Tile]) -> Village? {
```
Comment (justindomingue) [inline Warfare/Warfare/Map.swift:180]: what's the point of this function? just call `.owner` on any tile in the set.

File: Warfare/Warfare/GameEngine.swift (hunk 2 above)
Comment (danhp) [inline Warfare/Warfare/GameEngine.swift]: nah, That's clear village

File: Warfare/Warfare/Map.swift (hunk 5 above)
Comment (danhp) [inline Warfare/Warfare/Map.swift:180]: When you invade a region you can split it into a region where the owner isn't in it.
This checks if the owner is there or not.

**Comment by justindomingue** [comment]
Ready to merge

**Comment by danhp** [comment]
Should close #15

File: Warfare/Warfare/GameEngine.swift [hunk 6]
```diff
@@ -10,13 +10,413 @@ class GameEngine {
...
     init() { }
-    
+
     func newGame() {
         self.loadMap(number: "1")
     }
 
+    // MARK: - Operations
+
+    // Should get called once a cycle (after all player have played their turn
```
Comment (justindomingue) [inline Warfare/Warfare/GameEngine.swift:22]: missing ending paren ;-)

File: Warfare/Warfare/GameEngine.swift [hunk 7]
```diff
@@ -10,13 +10,413 @@ class GameEngine {
...
+
+            // Only consider tiles with trees
+            if t.land != .Tree { continue }
+
+            for n in self.map.neighbors(tile: t) {
+                if contains(seen, { $0 === n }) { continue }
+                seen.append(n)
+
+                if n.isGrowable() {
+                    let random = Int(arc4random_uniform(2))
```
Comment (justindomingue) [inline Warfare/Warfare/GameEngine.swift:40]: shouldn't it be arc4random_uniform(5) so 20% chance?

Comment (danhp) [inline Warfare/Warfare/GameEngine.swift:40]: it's 50% on one random neigbour

Comment (justindomingue) [inline Warfare/Warfare/GameEngine.swift:40]: I'm not sure how "random" that neighbour really is provided that map.neighbors() returns the neighbours from top left to down left clockwise. But I get that you only check unseen tiles.

**Pull request merged**